pyrad/packet.py

```diff
@@ -177,8 +179,21 @@ def verify_message_authenticator(self, secret=None,
         else:
             key = self.secret
 
-        self['Message-Authenticator'] = 16 * six.b('\00')
-        attr = self._PktEncodeAttributes()
+        # If there's a raw packet, use that to calculate the expected
+        # Message-Authenticator. While the Packet class keeps multiple
+        # instances of an attribute grouped together in the attribute list,
+        # other applications may not. Using _PktEncodeAttributes to get
+        # the attributes could therefore end up changing the attribute order
+        # because of the grouping Packet does, which would cause
+        # Message-Authenticator verification to fail. Using the raw packet
+        # instead, if present, ensures the verification is done using the
+        # attributes exactly as sent.
+        if self.raw_packet:
+            attr = self.raw_packet[20:]
+            attr = attr.replace(prev_ma[0], 16 * six.b('\00'))
+        else:
+            self['Message-Authenticator'] = 16 * six.b('\00')
+            attr = self._PktEncodeAttributes()
```

Comment:
This is the bug fix. This is essentially the same bug as in #140 , but for Message-Authenticator verification instead of authenticator verification. The fix here is to again use the raw_packet, but in this method we're not guaranteed it's present. The assumption here is that if a raw_packet is present, this Packet was instantiated using raw bytes received over the wire, and those bytes should be used for verification of the Message-Authenticator.